Add v23.1.0-alpha.4 release notes (#16366)
cockroachdb/docs: _includes/releases/v23.1/v23.1.0-alpha.4.md

- Users can now go to the next page of results when there are >20 active `stmts` or `txns` in the active exec pages. [#97122][#97122]
- Since 22.1 when rangefeed enablement overrides in span configs were introduced, rangefeed requests that reached spans outside the range would not cause range cache invalidation due to the setting being checked first, thus requests could repeatedly hit the same incorrect range, causing errors until cache invalidation or node restart. This fix correctly checks that the span is within the range prior to checking the enablement settings, thus invalidating the cache when a request reaches an incorrect range and causing subsequent requests to successfully reach the correct range. [#97212][#97212]
- A bug has been fixed that caused errors when creating multiple user-defined functions with the same name and different argument types in the same type family. For example, it was impossible to create both functions `f(i INT2)` and `f(INT4)`. [#96481][#96481]

> - The following spammy log message was removed: > lease [...] expired before being followed by lease [...]; foreground traffic may have been impacted [#97358][#97358]

Reviewer: I would rewrite to avoid slang like "spammy" and to use code font for the literal message that is no longer logged.

> - Previously, `ALTER TABLE ... INJECT STATISTICS` command would fail if a column with COLLATED STRING type had histograms to be injected, and this is now fixed. The bug has been present since at least 21.2. [#96695][#96695]

Reviewer: I would rewrite this as ".... no longer fails", but I'm confused about what used to fail. Injecting histograms?

> - Fixed a bug in the query engine that could cause incorrect results in some cases when a zigzag join was planned. The bug could occur when the two indexes used for the zigzag join had a suffix of matching columns but with different directions. For example, planning a zigzag join with `INDEX(a ASC, b ASC)` and `INDEX(c ASC, b DESC)` could cause incorrect results. This bug has existed since at least v19.1. It is now fixed, because the optimizer will no longer plan a zigzag join in such cases. [#97151][#97151]

Reviewer: How about "Planning a zigzag join query no longer fails when....."?

<h3 id="v23-1-0-alpha-4-performance-improvements">Performance improvements</h3>

> - The optimizer now plans inverted index scans for queries that filter by JSON fetch value operators `(->)`, consisting of both string and integer index values, alongside the IN operator, e.g, `json_col->0 IN ('1', '2')` and `json_col->'a' IN ('1', '2')`. [#96471][#96471]

Reviewer: ```suggestion
- The optimizer now plans inverted index scans for queries that filter by JSON fetch value operators `(->)` that contain both a string and an integer index value after the IN operator. For example the optimizer plans inverted index scans for the following filters:  `json_col->0 IN ('1', '2')`,  `json_col->'a' IN ('1', '2')`. [#96471][#96471]
```

<h3 id="v23-1-0-alpha-4-build-changes">Build changes</h3>

> - As a part of this change we start publishing a FIPS compliant tarball and docker image for the Linux x86_64 platform. The build uses OpenSSL libraries for crypto operations by `dlopen`ing the corresponding dynamic libraries. [#96107][#96107]

Reviewer: ```suggestion
- Starting with CockroachDB v23.1.0-alpha.4, a FIPS compliant tarball and Docker image are produced for the Linux x86_64 platform. The tarball uses OpenSSL libraries for crypto operations by dynamically loading the corresponding FIPS-validated dynamic libraries. The Docker image comes with the FIPS-validated OpenSSL library pre-installed.[#96107][#96107]
```
PTAL @rail

<h3 id="v23-1-0-alpha-4-miscellaneous">Miscellaneous</h3>

> <h4 id="v23-1-0-alpha-4-missing-category">Missing category</h4>

Reviewer: I'd remove this line.

-  [#97229][#97229]
- Fix a bug in which `RESTORE`, `BACKUP`, and `IMPORT` jobs would fail if the coordinator node of the job was drained. [#97033][#97033]

<h4 id="v23-1-0-alpha-4-changes-without-release-note-annotation">Changes without release note annotation</h4>

> - [#97344][#97344] [42a849c67][42a849c67] ui: fix jobs page requests (Xin Hao Zhang <xzhang@cockroachlabs.com>)

Reviewer: If this should be included, it could use a rewrite.

> - [#97171][#97171] [7ff5bfd74][7ff5bfd74] kvserver: print command ID for raft log entries (Tobias Grieger <tobias@tkschmidt.me>)

Reviewer: Does this actually need a release note?

> <h3 id="v23-1-0-alpha-4-doc-updates">Doc updates</h3>

Reviewer: This section can be removed.

{% comment %}Docs team: Please add these manually.{% endcomment %}

<div class="release-note-contributors" markdown="1">

<h3 id="v23-1-0-alpha-4-contributors">Contributors</h3>

This release includes 124 merged PRs by 48 authors.
We would like to thank the following contributors from the CockroachDB community:

> - DQ (first-time contributor)

Reviewer: I think all of these are CRL employees -- verify and remove line 89-95 if so.

<h3 id="v23-1-0-alpha-4-{{-site.data.products.enterprise-}}-edition-changes">{{ site.data.products.enterprise }} edition changes</h3>

> - Deprecate and replace some of the changefeed specific functions made available in preview mode in 22.2 release.   While these functions are deprecated, old changefeed transformations should continue to function properly. Customers are encouraged to closely monitor their changefeeds created during upgrade. While effort was made to maintain backward compatibility, the updated changefeed transformation may produce slightly different output (different column names, etc). [#96295][#96295]

Reviewer: ```suggestion
- Some of the transformations specific to changefeeds have been deprecated and replaced. These functions were released in limited  access in v22.2..  Deprecated changefeed transformations continue to function. Closely monitor changefeeds that are created during upgrade. While effort was made to maintain backward compatibility, the updated changefeed transformation may produce slightly different output, such as different column names. [#96295][#96295]
```

- Add support for implicit authentication to Azure Storage and KMS. [#96825][#96825]
- Add support for `CREATE EXTERNAL CONNECTION ... AS "postgresql://"` or `"postgres://"`. These external connections may be specified as the source in streaming replication. [#96551][#96551]
- Add support for referencing [user defined functions (UDFs)](../{{site.versions["stable"]}}/user-defined-functions.html) from other objects. Backup and restore operations can now read and write UDF descriptors. [#97038][#97038]

<h3 id="v23-1-0-alpha-4-sql-language-changes">SQL language changes</h3>

- UDFs can now return a set of results by setting the return type to `SETOF`. [#96698][#96698]
- UDFs with implicit record return types will return an error when called if the return type has been altered and is no longer compatible with the body of the UDF. [#96696][#96696]
- Implemented the `COPY ... TO STDOUT` statement, which allows exporting a table or arbitrary query using the postgres wire-compatible format. Only text & CSV format is supported. [#94408][#94408]

> - Add the read-only "ssl" session variable. It returns "off" if the server was started in insecure mode (which disables TLS), or "on" otherwise. This is based on the Postgres variable of the same name. [#97257][#97257]

Reviewer: I don't think a session variable "returns" anything -- reword that it is set to "off" or "on"?

<h3 id="v23-1-0-alpha-4-{{-site.data.products.enterprise-}}-edition-changes">{{ site.data.products.enterprise }} edition changes</h3>

> - Some of the transformations specific to changefeeds have been deprecated and replaced. These functions were released in limited  access in v22.2..  Deprecated changefeed transformations continue to function. Closely monitor changefeeds that are created during upgrade. While effort was made to maintain backward compatibility, the updated changefeed transformation may produce slightly different output, such as different column names. [#96295][#96295]

Reviewer: ```suggestion
- Some of the transformations specific to changefeeds have been deprecated and replaced. These functions were released in limited access in v22.2.  Deprecated changefeed transformations continue to function. Closely monitor changefeeds that are created during upgrade. While effort was made to maintain backward compatibility, the updated changefeed transformation may produce slightly different output, such as different column names. [#96295][#96295]
```

<h3 id="v23-1-0-alpha-4-operational-changes">Operational changes</h3>

- Add `COCKROACH_RAFT_LOG_QUEUE_CONCURRENCY` env var which controls the number of parallel workers doing Raft log truncations. It can be used to make the in-memory log truncations more agressive and reduce the amount of Raft log data flushed to disk. [#97029][#97029]
- The new timeseries metric `storage.keys.tombstone.count` shows the current count of point and range deletion tombstones across the storage engine. [#97306][#97306]
- The value of `kv.range_split.load_cpu_threshold` controls the CPU per-second threshold above which a range will be split. Its default value has been increased from `250ms` to `500ms`, based on performance experiments. [#97113][#97113]
- CPU balancing is enabled as the default load based rebalancing objective. This can be reverted by setting `kv.allocator.load_based_rebalancing.objective` to `qps`. [#97424][#97424]

<h3 id="v23-1-0-alpha-4-db-console-changes">DB Console changes</h3>

- Add columns p50, p90, p99, max and min latency for Statement table on SQL Activity page. [#97082][#97082]

> - Show a warning for Statement Insights when we reach a "max size exceed" error from the sql api. [#97153][#97153]

Reviewer: ```suggestion
- Show a warning for Statement Insights when the SQL API returns a "max size exceed" error. [#97153][#97153]
```

> - Show a warning for Transaction Insights when we reach a "max size exceed" error from the sql api. [#97277][#97277]

Reviewer: ```suggestion
- Show a warning for Transaction Insights when the SQL API returns a "max size exceed" error. [#97277][#97277]
```

- Fixed a bug where the `AS OF SYSTEM TIME` clause was handled incorrectly in an implicit transaction that had multiple statements. [#97063][#97063]
- This patch fixes asymmetric typing of `>` and `<` expressions which may cause erroring of expressions which are legal. [#97022][#97022]
- This patch fixes possible internal errors in `SOME`, `ANY` and `ALL` expressions of the form: `expr > SOME(expr1, expr2, expr3...)` [#97022][#97022]
- Fixed the syntax error for `SELECT ... QUERY` (without `AS`) statement. [#97041][#97041]
- Decommissions that would violate constraints set on a subset of replicas for a range (e.g. `num_replicas = 3, <constraint>: 1`) will no longer be able to execute, respecting constraints during and after the decommission. [#94810][#94810]
- Adjusted the size reported for `TIME`, `TIMETZ`, `TIMESTAMP`, `TIMESTAMPTZ`, and `DATE` types in the pgwire protocol. [#97145][#97145]
- The ParameterStatus message is now only sent over the pgwire protocol if the value of the parameter changed. (The parameters that are sent this way are timezone, intervalstyle, datestyle, is_superuser, and application_name.) [#97145][#97145]

> - Users can now go to the next page of results when there are >20 active statements or transactions in the Active Execution page. [#97122][#97122]

Reviewer: ```suggestion
- Users can now go to the next page of results when there are more than 20 active statements or transactions in the Active Execution page. [#97122][#97122]
```

<h3 id="v23-1-0-alpha-4-operational-changes">Operational changes</h3>

> - Add COCKROACH_RAFT_LOG_QUEUE_CONCURRENCY env var which controls the number of parallel workers doing Raft log truncations. It can be used to make the in-memory log truncations more agressive and reduce the amount of Raft log data flushed to disk. [#97029][#97029]

Reviewer: I think I had a suggested edit here but it is now hidden if so. The environment variable should be in code font, and "env var" should be changed to "environment variable".
